11x-new-kafka-roller.md

```diff
@@ -0,0 +1,227 @@
+
+
+## Motivation
+
+Strimzi users have been reporting some of the issues mentioned above and would benefit from a new KafkaRoller that is designed to address the shortcomings of the current KafkaRoller.
+
+The current code for KafkaRoller does not easily allow growth and adding new functionality such as batch rolling due to its complexity and nested conditions. It makes it challenging for users to debug and understand actions taken on their brokers when things go wrong and configure it correctly for their use cases. It is also not particularly easy to unit test which results in insufficient test coverage for many edge cases, making it challenging to refactor safely. 
+
+As mentioned in `Known Issues` section, the current KafkaRoller still needs various changes and potentially more as we get more experience with KRaft and discover more issues. Adding these non trivial changes to a component that is very complex and hard to reason, is expensive and poses potential risks of introducing bugs because of tightly coupled logics and lack of testability.
+
+Given the number of new features and changes related to KRaft, it is easiest to rewrite it from scratch rather than refactoring the existing component.  Rewriting with a more structured design, the process for evaluating nodes in various states would become more defined and easier to follow.
+
+## Proposal
+
+The objective of this proposal is to introduce a new KafkaRoller with more structured design resembling a finite state machine. The finite station machine design will make it easier to introduce the following features:
+- Ability to roll brokers in parallel to speed up updates in large clusters.
+- After restarting a broker, allow it to lead the partitions it is the preferred leader for. This will help reducing the impact on clients. 
+- Ability to restart controller nodes in parallel when needed, to help recovering controller qourum.
+- Add configurable wait time between restarts to allow slowing down rolling update.
+
+The new roller will have the same behaviour of the current roller but with the additional features above, however, the implementation will be different following the finite state machine design. 
+
+In this proposal term `node` refers to Kafka process that is either controller, broker or combined. When the term `pod` is used, it refers to the Kubernetes `Pod` resource where one of these Kafka processes is running inside.
+
+## State Machine
+
+Kafka nodes will be observed to transition into states and based on these states, actions will be taken on the nodes, which would result in another state transitions. This process will be executed on the nodes in a specific order and will be repeated for each node until the desired state or the maximum number of attempts is reached. The order will depend on which role the node has and what the currently observed state. This will be similar to the order the current roller executes:
+1. unready controllers
```

Comment:
Is this the same order that we have today? If not we should be clear how it's different, if it's the same then maybe it doesn't need to be called out?